### DISCUSSION TITLE
Change Hover Color for all Controls

### DISCUSSION THREAD

**Discussion by ShrutiJaiswal1494** [AvaloniaUI/Avalonia#6369] [Q&A] [answered]
Hi,

I want to change the default hover color of all the controls in my application. I am using the `Fluent` theme for DataGrid and `FluentLight` theme for my application. Is there a way to set the color through code? Tried looking through the web but cannot find anything relevant. Any help is much appreciated. Thanks in advance..!!

**Comment by maxkatz6** [comment, marked as answer]
There is a file https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Themes.Fluent/Accents/AccentColors.xaml with accent colors. You can redefine them or replace whole file.

And there are specific controls color resources https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Themes.Fluent/Accents/FluentControlResourcesDark.xaml . You also can redefine them or replace whole file.

By redefining I mean creating your own style-resources file with putting brushes with the same key but your color value in there. This approach is recommended and simpler, but there might be issues with some specific controls (new value can be ignored).

By replacing whole file I mean replacing FluentTheme style include from App.xaml and instead including theme parts manually. I.e. if you open FluentDark content - https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Themes.Fluent/FluentDark.xaml , you will see it's simple and you can copy that into your application instead of ```<FluentTheme />```. And instead of default AccentColors.xaml you can put there your own one.

**Reply by ShrutiJaiswal1494** [reply]
@maxkatz6 I tried redefining the colors in My `App.xaml` file like below but it's not getting changed. Am I doing something Wrong?

```
 <Application.Resources>
      <Color x:Key="SystemAccentColor">#FFFFE0</Color> 
  </Application.Resources>
```

**Reply by maxkatz6** [reply]
@ShrutiJaiswal1494 can you show full App.xaml file listing? Order is important there.
I also would create separated axaml styles file and include it using StyleInclude line after FluentTheme.

**Reply by ShrutiJaiswal1494** [reply]
@maxkatz6 Thanks for replying. My app file looks like this

```
<Application xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:local="clr-namespace:Test"
             xmlns:s="clr-namespace:System;assembly=mscorlib"
             x:Class="Test.App">
  <Application.DataTemplates>
    <local:ViewLocator/>
  </Application.DataTemplates>
  <Application.Resources>
      <Color x:Key="SystemAccentColor">#FFFFE0</Color>
    <DrawingGroup x:Key="Eye">
      <GeometryDrawing Brush="{DynamicResource ImageColor}" Geometry="F1M9,5.0625C10.5533008575439,5.0625,11.8125,6.32169914245605,11.8125,7.875C11.8125,8.87980937957764 11.2764406204224,9.8082914352417 10.40625,10.3106966018677 9.53605937957764,10.8131008148193 8.46394062042236,10.8131008148193 7.59374952316284,10.3106966018677C6.72355937957764,9.8082914352417,6.1875,8.87980937957764,6.1875,7.875C6.1875,6.32169914245605,7.44669914245605,5.0625,9,5.0625z M9,3.9375C6.82537889480591,3.9375,5.0625,5.70037889480591,5.0625,7.875C5.0625,10.0496215820313,6.82537889480591,11.8125,9,11.8125C10.4067335128784,11.8125 11.7066087722778,11.0620164871216 12.4099750518799,9.84375 13.1133413314819,8.62548351287842 13.1133413314819,7.12451601028442 12.4099750518799,5.90624952316284C11.7066078186035,4.68798303604126,10.4067335128784,3.9375,9,3.9375z M9,1.6875C14.625,1.6875 18,7.875 18,7.875 18,7.875 14.625,14.0625 9,14.0625 3.375,14.0625 0,7.875 0,7.875 0,7.875 3.375,1.6875 9,1.6875z" />
    </DrawingGroup>
    <s:Boolean x:Key="True">True</s:Boolean>
    <s:Boolean x:Key="False">False</s:Boolean>
  </Application.Resources>
  <Application.Styles>
    <Style Selector="GridSplitter.Horizontal">
      <Setter Property="Template">
        <Setter.Value>
          <ControlTemplate TargetType="{x:Type GridSplitter}">
            <Grid>
              <TextBlock HorizontalAlignment="Center" VerticalAlignment="Top" Text="····" Foreground="Gray" FontSize="26" FontWeight="Bold"/>
              <Rectangle Fill="#00FFFFFF" />
            </Grid>
          </ControlTemplate>
        </Setter.Value>
      </Setter>
    </Style>
  </Application.Styles>
</Application>

```

**Reply by ShrutiJaiswal1494** [reply]
Also the `FluentTheme` I am adding from code behind to switch between dark and light mode based on the User's choice So I am not sure where exactly should I include the accent color styles now.

**Reply by maxkatz6** [reply]
You need to ensure that your resource is defined AFTER fluent theme default one. In any way you could do it. Otherwise, it will be just overwritten.
Above I left my proposal how to do it with separated axaml styles file, that you can reference with StyleInclude. You also can add it from the code behind after fluent theme to the same styles collection. Or you can insert fluent theme at 0 index before any custom styles.

**Reply by ShrutiJaiswal1494** [reply]
@maxkatz6  As you suggested I am trying to create a separate Styles file and include it through `StyleInclue` but I am not sure what should I give for the URI. I tried using the local path for the styles file but doesn't work it seems.

![image](https://user-images.githubusercontent.com/79670853/128765967-d6f2639f-b881-4274-91af-f17fdb168b3b.png)

```
private static readonly StyleInclude s_DataGridFluent = new StyleInclude(new Uri("avares://ControlCatalog/Styles"))
        {
            Source = new Uri("avares://Avalonia.Controls.DataGrid/Themes/Fluent.xaml")
        };

private static readonly StyleInclude s_ColorStyles = new(new Uri("C:/Users/shjaiswal/source/repos/MyRepo/MyProject/TestUI"))
        {
            Source = new Uri("/Styles/Colors.axaml")
        };

        public static Styles FluentLight = new()
        {
            new StyleInclude(new Uri("avares://ControlCatalog/Styles"))
            {
                Source = new Uri("avares://Avalonia.Themes.Fluent/FluentLight.xaml")
            },
            s_DataGridFluent,
            s_ColorStyles
        };
```

**Reply by maxkatz6** [reply]
Path is relative to project root:
```
new(new Uri("avares://Test/Styles"))
        {
            Source = new Uri("avares://Test/Styles/Colors.axaml")
        };
```
Or maybe it will work too:
```
new(new Uri("avares://Test"))
        {
            Source = new Uri("/Styles/Colors.axaml")
        };
```
"Test" is project name.

**Reply by ShrutiJaiswal1494** [reply]
Tried both. The first one is throwing a null pointer exception  in `Style.cs`
![image](https://user-images.githubusercontent.com/79670853/128768517-24b6d0aa-7189-4c8a-bd34-858ffdb681bf.png)


And the second one throws a `System.TypeInitializationException`

**Reply by maxkatz6** [reply]
can you share your test project?

**Reply by ShrutiJaiswal1494** [reply]
Got the first example working. Thank you so much :)
